src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs

```diff
@@ -414,21 +414,78 @@ internal ComponentInfo(Type interfaceType, string kind, Type argumentType, TlcMo
             ctor = null;
             create = null;
             requireEnvironment = false;
+            bool requireEnvironmentCtor = false;
+            bool requireEnvironmentCreate = false;
             var parmTypesWithEnv = Utils.Concat(new Type[1] { typeof(IHostEnvironment) }, parmTypes);
+
             if (Utils.Size(parmTypes) == 0 && (getter = FindInstanceGetter(instType, loaderType)) != null)
                 return true;
-            if (instType.IsAssignableFrom(loaderType) && (ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypes ?? Type.EmptyTypes, null)) != null)
-                return true;
-            if (instType.IsAssignableFrom(loaderType) && (ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypesWithEnv ?? Type.EmptyTypes, null)) != null)
+
+            // Find both 'ctor' and 'create' methods if available
+            if (instType.IsAssignableFrom(loaderType))
             {
-                requireEnvironment = true;
-                return true;
+                ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypes ?? Type.EmptyTypes, null);
+                if (ctor == null)
+                {
+                    ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypesWithEnv ?? Type.EmptyTypes, null);
+                    if (ctor != null)
+                        requireEnvironmentCtor = true;
+                }
             }
-            if ((create = FindCreateMethod(instType, loaderType, parmTypes ?? Type.EmptyTypes)) != null)
+
+            create = FindCreateMethod(instType, loaderType, parmTypes ?? Type.EmptyTypes);
+            if (create == null)
+            {
+                create = FindCreateMethod(instType, loaderType, parmTypesWithEnv ?? Type.EmptyTypes);
+                if (create != null)
+                    requireEnvironmentCreate = true;
+            }
+
+            // If both 'ctor' and 'create' methods were found
+            // Choose the one that is public, if both are non-public choose the one that isn't private
+            // if they have the same visibility, then throw an exception, since this shouldn't happen.
+            if(ctor != null && create != null)
+            {
```

Comment:
I don't really like this nested if's, I am not sure if they're legible enough, and it is ambiguous what should happen if there's a 'protected' create or constructor method (which, I believe, never happens in the codebase...). Still, this gets the job done.

I can think of a couple of ways of making this, but not sure if they would be more legible. Please, let me know if I should rewrite this in another way.

Reply:
So I've changed this nested if's in a new iteration ([see this comment](https://github.com/dotnet/machinelearning/pull/4485#discussion_r347688260)) but I am not sure if I prefer the nested of's or the new solution.